src/data/matrix/basic.lean

```diff
@@ -77,6 +77,20 @@ instance [has_neg α] : has_neg (matrix m n α) := pi.has_neg
 instance [has_sub α] : has_sub (matrix m n α) := pi.has_sub
 instance [add_group α] : add_group (matrix m n α) := pi.add_group
 instance [add_comm_group α] : add_comm_group (matrix m n α) := pi.add_comm_group
+instance [unique α] : unique (matrix m n α) := pi.unique
+instance [subsingleton α] : subsingleton (matrix m n α) := pi.subsingleton
+instance [nonempty m] [nonempty n] [nontrivial α] : nontrivial (matrix m n α) :=
+function.nontrivial
+
+/--
+A 1 × 1 `matrix` can be considered as a value of the underlying scalar.
```

Comment:
I'm not convinced there is a case where you actually want this coercion. Can you think of one?

Reply:
I agree that I would like to have a bit more evidence for why we want this coe.
The rest is uncontroversial :wink:

Reply:
One can then have something like this:
```
def dot_product' [has_mul α] [add_comm_monoid α] (v w : m → α) : α :=
matrix.mul (row v) (col w)
```

Of course, `matrix.mul` is defined using `dot_product` currently. But this sort of coercion to calculate the scalar value from row and column vectors is (ab)used often. So this alternative `dot_product'` definition isn't needed -- just showing how someone might try to phrase such usage.

Reply:
I remain skeptical based on that example. :-)

Reply:
Removed.